constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala

```diff
@@ -0,0 +1,120 @@
+ * limitations under the License.
+ */
+
+package de.heikoseeberger.constructr.coordination
+
+import akka.http.scaladsl.client.RequestBuilding.{ Get, Put }
+import akka.http.scaladsl.model.StatusCodes.{ NotFound, OK }
+import akka.http.scaladsl.model.{ HttpRequest, HttpResponse, ResponseEntity, Uri }
+import akka.http.scaladsl.unmarshalling.Unmarshal
+import akka.stream.Materializer
+import scala.concurrent.duration.Duration
+import scala.concurrent.{ ExecutionContext, Future }
+
+final class ConsulCoordination(prefix: String, clusterName: String, host: String, port: Int, send: HttpRequest => Future[HttpResponse])
+    extends Coordination[Coordination.Backend.Consul.type](prefix, clusterName, host, port, send) {
+  import Coordination._
+
+  private val v1Uri = Uri(s"http://$host:$port/v1")
+
+  private val kvUri = Uri(s"http://$host:$port/v1/kv")
+
+  private val baseUri = kvUri.withPath(kvUri.path / "constructr" / prefix / clusterName)
+
+  private val nodesUri = baseUri.withPath(baseUri.path / "nodes")
+
+  private val createSessionUri = v1Uri.withPath(v1Uri.path / "session" / "create")
+
+  override def getNodes[A: AddressSerialization]()(implicit ec: ExecutionContext, mat: Materializer): Future[List[A]] = {
+    def unmarshalNodes(entity: ResponseEntity) = {
+      def toNodes(s: String) = {
+        import rapture.json._
+        import rapture.json.jsonBackends.spray._
+        def jsonToNode(json: Json) = {
+          implicitly[AddressSerialization[A]].fromBytes(decode(json.Key.as[String].substring(s"constructr/$prefix/$clusterName/nodes/".length)))
+        }
+        Json.parse(s).as[List[Json]].map(jsonToNode)
+      }
+      Unmarshal(entity).to[String].map(toNodes)
+    }
+    send(Get(nodesUri.withQuery(Uri.Query("recurse")))).flatMap {
+      case HttpResponse(OK, _, entity, _)       => unmarshalNodes(entity)
+      case HttpResponse(NotFound, _, entity, _) => ignore(entity).map(_ => Nil)
+      case HttpResponse(other, _, entity, _)    => ignore(entity).map(_ => throw UnexpectedStatusCode(other))
+    }
+  }
+
+  override def lock(ttl: Duration)(implicit ec: ExecutionContext, mat: Materializer): Future[LockResult] = {
+    // TODO Make idempotent, e.g. by first GETting and comparing the value, just PUTting, if not there with value = self
+    val uri = baseUri
+      .withPath(baseUri.path / "lock")
+
+    createSession(ttl)
+      .flatMap(sessionId => send(Put(uri.withQuery(Uri.Query("acquire" -> sessionId)))))
+      .flatMap {
+        case HttpResponse(OK, _, entity, _) => {
+          Unmarshal(entity).to[String]
+            .map(_.toBoolean)
+            .map(locked => if (locked) LockResult.Success else LockResult.Failure)
+        }
+        case HttpResponse(other, _, entity, _) =>
+          ignore(entity).map(_ => throw UnexpectedStatusCode(other))
+      }
+  }
+
+  override def addSelf[A: AddressSerialization](self: A, ttl: Duration)(implicit ec: ExecutionContext, mat: Materializer) =
+    (for {
```

Comment:
This is hard to read. Use an intermediate result assigned to a val, e.g. `val responseAndSessionId = ...`, for the for-comprehension. Then call `flatMap` on that identifier.